### PR TITLE
New version of pyspark mnist kmeans example notebook

### PR DESCRIPTION
**Pull request by JonathanTaws**
Proposal for a new version and revision of the content for the pyspark mnist kmeans example notebook. 

Main changes:
- Included example of using Spark pipeline with hybrid or SageMaker-only pipelines
- Showed how to create custom SageMakerEstimator in SageMaker-Spark SDK
- Showcase re-use of endpoints, models, and training job data
- Enhance clean up to clean up models in Pipelines

I also have a few diagrams for these that I can include. As they are images, it might increase the load time of the Jupyter notebook browser however.

### CODE REVIEW

**Comment by andremoeller** [comment]
Thanks @JonathanTaws !

The content is excellent. It might make sense to split this up into multiple notebooks, since the running time for the whole notebook is pretty long: one for hybrid pipeline, one for SageMaker-only pipeline, one for custom SageMakerEstimator. (I think it makes sense to keep re-using existing endpoints in this one.).

I think it's fine to include the images in the notebooks. The effect of reasonably sized images on notebook load time should be negligible.

Thoughts @djarpin ?

**Comment by JonathanTaws** [comment]
I went ahead and split up the notebookes, and included some diagrams. Any thoughts @andremoeller and @djarpin ?

File: sagemaker-spark/pyspark_mnist/pyspark_mnist_custom_estimator.ipynb [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+    "version": 3
+   },
+   "file_extension": ".py",
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.6.4"
+  },
+  "notice": "Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.  Licensed under the Apache License, Version 2.0 (the \"License\"). You may not use this file except in compliance with the License. A copy of the License is located at http://aws.amazon.com/apache2.0/ or in the \"license\" file accompanying this file. This file is distributed on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License."
```
Comment (djarpin) [inline sagemaker-spark/pyspark_mnist/pyspark_mnist_custom_estimator.ipynb]: Can you change to 2018?

File: sagemaker-spark/pyspark_mnist/pyspark_mnist_custom_estimator.ipynb [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+   "metadata": {},
+   "source": [
+    "![Custom estimator on SageMaker](img/sagemaker-spark-custom-architecture.png)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's train this estimator by calling fit on it with the training data."
```
Comment (djarpin) [inline sagemaker-spark/pyspark_mnist/pyspark_mnist_custom_estimator.ipynb]: Can you call out this one and the following cells that will take a bit of time to run?  Thanks.

**Comment by lynnlangit** [comment]
I am really looking forward to trying out this set of sample notebooks after the PR is merged.

File: sagemaker-spark/pyspark_mnist/pyspark_mnist_pca_mllib_kmeans.ipynb [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    "\n",
+    "As we are interested in clustering the images of digits, the number of pixels represents the feature vector, while the number of classes represents the number of clusters we want to find. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Create a hybrid pipeline with Spark PCA and SageMaker K-Means\n",
+    "To perform the clustering task, we will first running PCA on our feature vector, reducing it to 50 features. Then, we can use K-Means on the result of PCA to apply the final clustering. We will create a **Pipeline** consisting of 2 stages: the PCA stage, and the K-Means stage. \n",
```
[the remark this answers is not in the document]
Comment (JonathanTaws) [reply sagemaker-spark/pyspark_mnist/pyspark_mnist_pca_mllib_kmeans.ipynb]: Good point. Would a sentence such as the following capture this? 
" The use of Spark MLLib PCA in this notebook is meant to showcase how you can use different pre-processting steps, ranging from data transformers to algorithms, with tools such as Spark MLLib that are well suited for data pre-processing. You can then use SageMaker algorithms and features through the SageMaker-Spark SDK. Here in our case, PCA is in charge of reducing the feature vector as a pre-processing step, and K-Means responsible for clustering the data. "

**Comment by JonathanTaws** [comment]
Thanks @djarpin for looking into this, I've taken into account all changes and they have been pushed into the PR. 
Let's review exactly how we want to frame the PCA on Spark vs. PCA on SageMaker part.

File: sagemaker-spark/pyspark_mnist/pyspark_mnist_pca_mllib_kmeans.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (djarpin) [reply sagemaker-spark/pyspark_mnist/pyspark_mnist_pca_mllib_kmeans.ipynb]: This sounds good to me.  Let's go with it.  Thanks.

**Comment by JonathanTaws** [comment]
Awesome! In that case we're good to go if there's nothing else to address.

Le lun. 30 avr. 2018 à 09:15, David Arpin <notifications@github.com> a
écrit :

**Pull request merged**